### ISSUE TITLE
Can't receive an answer from RPC request to Neovim created as child process

### ISSUE THREAD

**Issue by echasnovski** [neovim/neovim#21630]
### Neovim version (nvim -v)

NVIM v0.9.0-dev-598+g3a519d86b

### Vim (not Nvim) behaves the same?

No (doesn't have this functionality)

### Operating system/version

EndeavourOS Linux x86_64 (6.1.1-arch1-1)

### Terminal name/version

st-0.9

### $TERM environment variable

st-256color

### Installation

appimage

### How to reproduce the issue

1. Create config file (called here 'init-nightly-job.lua') with the following content:
```lua
_G.new_child_neovim = function(nvim_executable)
  local stdin = vim.loop.new_pipe()
  local stdout = vim.loop.new_pipe()
  local stderr = vim.loop.new_pipe()

  vim.loop.spawn(nvim_executable or 'nvim', {
    stdio = { stdin, stdout, stderr },
    args = { '--clean', '--listen', '/tmp/nvim.pipe' },
  }, function() end)
end
```
2. Open Neovim nightly with `nvim_nightly_g598 -u init-nightly-job.lua` (assuming `nvim_nightly_g598` is an executable for used Neovim nightly release).
3. `:lua new_child_neovim('nvim_nightly_g598')` . This should open child Neovim process (can be confirmed by looking at list of system processes).
4. `:lua channel = vim.fn.sockconnect('pipe', '/tmp/nvim.pipe', { rpc = true })`. This should create a valid channel.
5. `:lua print(vim.rpcrequest(channel, 'nvim_get_mode'))` . This does not respond.

### Expected behavior

Can get responds from child Neovim process.

This works as expected in the following situations:
- With current stable Neovim 0.8.2 release:
    - `nvim_nightly_g598 -u init-nightly-job.lua`
    - `:lua new_child_neovim('nvim_0.8.2')` with `nvim_0.8.2` being Neovim 0.8.2 executable.
    - `:lua channel = vim.fn.sockconnect('pipe', '/tmp/nvim.pipe', { rpc = true })`
    - `:lua print(vim.rpcrequest(channel, 'nvim_get_mode'))`
- When child Neovim nightly is opened directly:
    - `nvim_nightly_g598 --clean --listen /tmp/nvim.pipe`
    - In another terminal:
        - `nvim_nightly_g598 -u NONE`
        - `:lua channel = vim.fn.sockconnect('pipe', '/tmp/nvim.pipe', { rpc = true })`
        - `:lua print(vim.rpcrequest(channel, 'nvim_get_mode'))`

Note: run `rm /tmp/nvim.pipe` to make it work after running Neovim nightly example.

### Actual behavior

Host process hangs, possibly waiting for answer. Only manually stopping child process can unhang it.

This seems to have been working around 4 days ago, so maybe related to recent work on TUI.

**Comment by echasnovski** [comment]
@bfredl, could you, please, share some insights into this? Is this indeed a regression or should this kind of task now be done differently? This disables my CI tests for nightly, so I am eager to get this working.

**Comment by bfredl** [comment]
I am working through all the reported regressions from #18375 .

**Comment by echasnovski** [comment]
So, this is indeed a regression, understood. Thanks! Will wait patiently.

**Comment by bfredl** [comment]
@echasnovski how is stdin `local stdin = vim.loop.new_pipe()` being used ? It seems nvim gets stuck trying to read stdin as a file.  If it is not used at all, I'd suggest add `--headless` flag to the cmdline to avoid this behavior.

**Comment by echasnovski** [comment]
> @echasnovski how is stdin `local stdin = vim.loop.new_pipe()` being used ? It seems nvim gets stuck trying to read stdin as a file. If it is not used at all, I'd suggest add `--headless` flag to the cmdline to avoid this behavior.

This approach was the result of not wanting to do `--headless` because it altered the behavior (not quite remember how; I think it was something with autocommands among other things). It is used to imitate real Neovim child process for testing and worked quite well for Neovim>=0.5.

The whole `stdin, stdout, stderr` was taken from `:h uv.spawn()` example in an attempt to create child Neovim process without dependencies. [Here is more context](https://github.com/echasnovski/mini.nvim/blob/main/lua/mini/test.lua#L1058).

**Comment by bfredl** [comment]
you could immediately close stdin after spawning as well.

**Comment by echasnovski** [comment]
> you could immediately close stdin after spawning as well.

Yes, adding `stdin:close()` at the end of function in original post seems to make it work.

Couple of follow-ups:
- Is it the long term ("correct") solution here or just a temporary one?
- Closing `stdout` and `stderr` seems to leave it working. Is it a good practice to also close them after spawn (in case only RPC functionality will be used for communicating)?

**Comment by bfredl** [comment]
this approach is expected to work in the long-term. But in addition, we want to simplify this use case by allow stdin from `vim.loop.spawn` to be used for RPC directly (but this will require more work in core).

**Comment by echasnovski** [comment]
> you could immediately close stdin after spawning as well.

Just FYI, this approach doesn't work for Neovim<0.8. Both immediate and not immediate closing works for Neovim 0.8.2. So some conditioning is required.

**Comment by echasnovski** [comment]
Fixed it on my end. The original issue is not an immediate problem for me now.

I am hesitant to close this myself, as it might still be an issue.

**Comment by echasnovski** [comment]
FYI, this seems to reproduce the same issue:

- `nvim -u NONE`
- `:lua channel = vim.fn.jobstart({ 'nvim', '--clean' }, { rpc = true })`
- `:lua vim.rpcrequest(channel, 'nvim_get_mode')`. It hangs current Neovim process on nightly (`NVIM v0.9.0-dev-825+g843c9025a`) and segfaults on Neovim 0.8.2.

**Comment by zeertzjq** [comment]
> `:lua channel = vim.fn.jobstart({ 'nvim', '--clean' }, { rpc = true })`

This is using stdio as rpc channel when it is not, so it is definitely not going to work. This works:
```vim
:lua channel = vim.fn.jobstart({ 'nvim', '--clean', '--headless', '--embed' }, { rpc = true })
```

**Comment by echasnovski** [comment]
A follow-up after testing a solution from #21886.

Using this way of opening child process solves the issue:

```lua
_G.new_child_neovim = function(nvim_executable)
  vim.fn.jobstart({ nvim_executable or 'nvim', '--clean', '--listen', '/tmp/nvim.pipe' }, { pty = true })
end
```

I don't have problem with this being closed, but I am still not sure if current situation is a desired behavior.
